{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/deepak.unni3/GIT/kgx/test/lib/python3.6/site-packages/cachier/mongo_core.py:24: UserWarning: Cachier warning: pymongo was not found. MongoDB cores will not work.\n",
      "  \"Cachier warning: pymongo was not found. MongoDB cores will not work.\")\n"
     ]
    }
   ],
   "source": [
    "from kgx import Transformer, NeoTransformer, PandasTransformer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Monarch-Lite\n",
    "\n",
    "# Credentials can be found from 'Registry of Biolink-compatible Neo4 instances' spreadsheet\n",
    "monarch_uri = ''\n",
    "monarch_username = ''\n",
    "monarch_password = ''\n",
    "\n",
    "# Initialize NeoTransformer\n",
    "monarch_lite_transformer = NeoTransformer(uri=monarch_uri, username=monarch_username, password=monarch_password)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define filters\n",
    "monarch_lite_transformer.set_filter(\"subject_category\", \"gene\")\n",
    "monarch_lite_transformer.set_filter(\"object_category\", \"disease\")\n",
    "monarch_lite_transformer.set_filter(\"edge_label\", \"contributes_to\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2018-06-08 18:40:06,916 root       DEBUG      MATCH (n:gene:disease ) RETURN n SKIP 0 LIMIT 1000;\n",
      "2018-06-08 18:40:07,018 root       DEBUG      MATCH (s:gene)-[p:contributes_to ]->(o:disease)\n",
      "            RETURN s,p,o\n",
      "            SKIP 0 LIMIT 1000;\n",
      "2018-06-08 18:40:07,620 root       DEBUG      MATCH (s:gene)-[p:contributes_to ]->(o:disease)\n",
      "            RETURN s,p,o\n",
      "            SKIP 1000 LIMIT 1000;\n",
      "2018-06-08 18:40:08,211 root       DEBUG      MATCH (s:gene)-[p:contributes_to ]->(o:disease)\n",
      "            RETURN s,p,o\n",
      "            SKIP 2000 LIMIT 1000;\n",
      "2018-06-08 18:40:08,831 root       DEBUG      MATCH (s:gene)-[p:contributes_to ]->(o:disease)\n",
      "            RETURN s,p,o\n",
      "            SKIP 3000 LIMIT 1000;\n",
      "2018-06-08 18:40:09,535 root       DEBUG      MATCH (s:gene)-[p:contributes_to ]->(o:disease)\n",
      "            RETURN s,p,o\n",
      "            SKIP 4000 LIMIT 1000;\n",
      "Total number of nodes: 5179\n",
      "Total number of edges: 5000\n"
     ]
    }
   ],
   "source": [
    "# Load nodes and edges from Monarch-Lite\n",
    "start = 0\n",
    "end = 5000\n",
    "monarch_lite_transformer.load(start=start, end=end)\n",
    "monarch_lite_transformer.report()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('NCBIGene:38903500',\n",
       "  {'iri': 'http://www.ncbi.nlm.nih.gov/gene/38903500',\n",
       "   'clique': ['NCBIGene:38903500'],\n",
       "   'name': 'KDSR',\n",
       "   'http://www.w3.org/2000/01/rdf-schema#label': 'KDSR',\n",
       "   'id': 'NCBIGene:38903500',\n",
       "   'label': 'KDSR',\n",
       "   'category': ['gene', 'sequence feature'],\n",
       "   'labels': ['Class', 'sequence_feature', 'gene', 'cliqueLeader', 'Node']}),\n",
       " ('OMIA:000939-9913',\n",
       "  {'iri': 'http://purl.obolibrary.org/obo/OMIA_000939-9913',\n",
       "   'clique': ['OMIA:000939-9913'],\n",
       "   'http://purl.org/dc/elements/1.1/description': 'By examining three comparative positional candidate genes from their mapping results (see Mapping section), Krebs et al. (2007) reported strong evidence that this disorder in cattle is due to a mutation in \"FVT1, encoding 3-ketodihydrosphingosine reductase, which catalyzes a crucial step in the glycosphingolipid metabolism\". Specifically, the mutation is \"a G-to-A missense mutation that changes Ala-175 to Thr\". Interestingly, this is not the same gene which, when mutated, results in SMA in humans.\\nIn their table of reduced-fertility haplotypes, Cole et al. (2014) list this KDSR mutation as being the causal mutation for haplotype BHM, which is located at 62,118,139��� 62,156,760bp on chromosome BTA24. [mol_gen]',\n",
       "   'name': 'Haplotype BHM',\n",
       "   'http://www.w3.org/2000/01/rdf-schema#label': 'Haplotype BHM',\n",
       "   'id': 'OMIA:000939-9913',\n",
       "   'label': 'Haplotype BHM',\n",
       "   'category': 'disease',\n",
       "   'labels': ['Class', 'Node', 'cliqueLeader', 'disease']}),\n",
       " ('NCBIGene:389756176',\n",
       "  {'iri': 'http://www.ncbi.nlm.nih.gov/gene/389756176',\n",
       "   'clique': ['NCBIGene:389756176'],\n",
       "   'name': 'TMEM154',\n",
       "   'http://www.w3.org/2000/01/rdf-schema#label': 'TMEM154',\n",
       "   'id': 'NCBIGene:389756176',\n",
       "   'label': 'TMEM154',\n",
       "   'category': ['gene', 'sequence feature'],\n",
       "   'labels': ['Class', 'sequence_feature', 'gene', 'cliqueLeader', 'Node']}),\n",
       " ('OMIA:001694-9940',\n",
       "  {'iri': 'http://purl.obolibrary.org/obo/OMIA_001694-9940',\n",
       "   'clique': ['OMIA:001694-9940'],\n",
       "   'http://purl.org/dc/elements/1.1/description': ['Following up the results of their GWAS described above, Heaton et al. (2012) showed that the two OAR17 SNPs are located in \"an ovine gene homologous to the human TMEM154 gene\". Subsequent sequencing and association analysis showed that \"Two TMEM154 haplotypes encoding glutamate (E) at position 35 were associated with infection.  . . .  The odds of infection for ewes with one copy of a full-length TMEM154 E35 allele were 28 times greater than the odds for those without (p-value<0.0001, 95% CI 5���1,100). In a combined analysis of nine cohorts with 2,705 sheep from Nebraska, Idaho, and Iowa, the relative risk of infection was 2.85 times greater for sheep with a full-length TMEM154 E35 allele (p-value<0.0001, 95% CI 2.36���3.43). Although rare, some sheep were homozygous for TMEM154 deletion mutations and remained uninfected despite a lifetime of significant exposure.\" Taken together, these results led to Heaton et al. (2012) concluding that \"TMEM154 may play a central role in ovine lentivirus infection and removing sheep with the most susceptible genotypes may help eradicate OPP and protect flocks from reinfection\". While the actual function of the membrane protein encoded by TMEM154 is still to be worked out, the apparent resistance of sheep homozygous for a mutant that results in no functional peptide strongly suggests a mechanism for resistance/susceptibility, namely that the lentivirus needs this peptide in order to infect cells.\\nFollowing up one of the most promising regions for control of infection from the GWAS of White et al. (2012), White et al. (2013) \"identified a small insertion/deletion variant near [5\\' of] ZNF389 [OVA20; g.29500068_29500069delAT; GenBank NC_019477.1] that showed consistent association with proviral concentration in three animal sets . . . The best estimate of proviral concentration by genotype, obtained from all 1310 OvLV-positive animals tested, showed insertion homozygotes had less than half the proviral concentration of other genotypes (P < 0.0001).  . . . To our knowledge, this is the first genetic variant consistently associated with host control of OvLV post-infection in multiple sheep flocks.\" [mol_gen]',\n",
       "    'As reported by Heaton et al. (2013), \"The first signs of disease typically appear after age two and often include the loss of body condition and indurative mastitis (hard udder). Disease progression is associated with severe clinical signs that include difficulty breathing, chronic wasting, loss of motor control, and arthritis.\" [clin_feat]'],\n",
       "   'name': 'Resistance/susceptibility to Visna/Maedi virus (VMV); Resist',\n",
       "   'http://www.w3.org/2000/01/rdf-schema#label': 'Resistance/susceptibility to Visna/Maedi virus (VMV); Resist',\n",
       "   'id': 'OMIA:001694-9940',\n",
       "   'label': 'Resistance/susceptibility to Visna/Maedi virus (VMV); Resist',\n",
       "   'category': 'disease',\n",
       "   'labels': ['Class', 'Node', 'cliqueLeader', 'disease']}),\n",
       " ('NCBIGene:200685841',\n",
       "  {'iri': 'http://www.ncbi.nlm.nih.gov/gene/200685841',\n",
       "   'clique': ['NCBIGene:200685841'],\n",
       "   'name': 'ABO',\n",
       "   'http://www.w3.org/2000/01/rdf-schema#label': 'ABO',\n",
       "   'id': 'NCBIGene:200685841',\n",
       "   'label': 'ABO',\n",
       "   'category': ['gene', 'sequence feature'],\n",
       "   'labels': ['Class', 'sequence_feature', 'gene', 'cliqueLeader', 'Node']}),\n",
       " ('OMIA:001089-9590',\n",
       "  {'iri': 'http://purl.obolibrary.org/obo/OMIA_001089-9590',\n",
       "   'clique': ['OMIA:001089-9590'],\n",
       "   'name': 'Blood group system ABO in siamang',\n",
       "   'http://www.w3.org/2000/01/rdf-schema#label': 'Blood group system ABO in siamang',\n",
       "   'id': 'OMIA:001089-9590',\n",
       "   'label': 'Blood group system ABO in siamang',\n",
       "   'category': 'disease',\n",
       "   'labels': ['Class', 'Node', 'cliqueLeader', 'disease']}),\n",
       " ('NCBIGene:403511',\n",
       "  {'iri': 'http://www.ncbi.nlm.nih.gov/gene/403511',\n",
       "   'clique': ['NCBIGene:403511', 'ENSEMBL:ENSCAFG00000014655'],\n",
       "   'synonym': ['transducin beta chain 3',\n",
       "    'guanine nucleotide-binding protein G(I)/G(S)/G(T) subunit beta-3',\n",
       "    'transducin beta-3-subunit',\n",
       "    'G protein subunit beta 3',\n",
       "    'GTP-binding protein beta-3 subunit short isoform'],\n",
       "   'http://www.geneontology.org/formats/oboInOwl#hasExactSynonym': 'G protein subunit beta 3',\n",
       "   'http://purl.org/dc/elements/1.1/description': 'G protein subunit beta 3',\n",
       "   'name': 'GNB3',\n",
       "   'http://www.w3.org/2000/01/rdf-schema#label': 'GNB3',\n",
       "   'http://www.geneontology.org/formats/oboInOwl#hasRelatedSynonym': ['transducin beta chain 3',\n",
       "    'guanine nucleotide-binding protein G(I)/G(S)/G(T) subunit beta-3',\n",
       "    'transducin beta-3-subunit',\n",
       "    'GTP-binding protein beta-3 subunit short isoform'],\n",
       "   'id': 'NCBIGene:403511',\n",
       "   'label': 'GNB3',\n",
       "   'category': ['gene', 'sequence feature'],\n",
       "   'labels': ['Class', 'sequence_feature', 'gene', 'cliqueLeader', 'Node']}),\n",
       " ('OMIA:001244-9615',\n",
       "  {'iri': 'http://purl.obolibrary.org/obo/OMIA_001244-9615',\n",
       "   'clique': ['OMIA:001244-9615'],\n",
       "   'name': 'Retinal degeneration II in dog',\n",
       "   'http://www.w3.org/2000/01/rdf-schema#label': 'Retinal degeneration II in dog',\n",
       "   'id': 'OMIA:001244-9615',\n",
       "   'label': 'Retinal degeneration II in dog',\n",
       "   'category': 'disease',\n",
       "   'labels': ['Class', 'Node', 'cliqueLeader', 'disease']}),\n",
       " ('NCBIGene:403972',\n",
       "  {'iri': 'http://www.ncbi.nlm.nih.gov/gene/403972',\n",
       "   'synonym': ['retinal degeneration slow protein',\n",
       "    'RDS',\n",
       "    'peripherin 2',\n",
       "    'peripherin protein',\n",
       "    'peripherin-2',\n",
       "    'retinal degeneration, slow',\n",
       "    'rds gene for peripherin'],\n",
       "   'clique': ['NCBIGene:403972', 'ENSEMBL:ENSCAFG00000001722'],\n",
       "   'http://www.geneontology.org/formats/oboInOwl#hasExactSynonym': 'peripherin 2',\n",
       "   'http://purl.org/dc/elements/1.1/description': 'peripherin 2',\n",
       "   'name': 'PRPH2',\n",
       "   'http://www.geneontology.org/formats/oboInOwl#hasRelatedSynonym': ['retinal degeneration slow protein',\n",
       "    'RDS',\n",
       "    'peripherin protein',\n",
       "    'peripherin-2',\n",
       "    'retinal degeneration, slow',\n",
       "    'rds gene for peripherin'],\n",
       "   'http://www.w3.org/2000/01/rdf-schema#label': 'PRPH2',\n",
       "   'id': 'NCBIGene:403972',\n",
       "   'label': 'PRPH2',\n",
       "   'category': ['gene', 'sequence feature'],\n",
       "   'labels': ['Class', 'sequence_feature', 'gene', 'cliqueLeader', 'Node']})]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show the first 10 nodes from the networkx graph\n",
    "monarch_lite_transformer.graph.nodes(data=True)[0:9]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('NCBIGene:38903500',\n",
       "  'OMIA:000939-9913',\n",
       "  {'is_defined_by': 'Monarch',\n",
       "   'provided_by': ['https://data.monarchinitiative.org/ttl/omia.ttl'],\n",
       "   'evidence': [],\n",
       "   'qualifier': 'direct',\n",
       "   'publications': [],\n",
       "   'relation': 'RO:0002326',\n",
       "   'subject': 'NCBIGene:38903500',\n",
       "   'object': 'OMIA:000939-9913',\n",
       "   'id': 3400163,\n",
       "   'type': 'contributes_to',\n",
       "   'predicate': 'contributes_to'}),\n",
       " ('NCBIGene:389756176',\n",
       "  'OMIA:001694-9940',\n",
       "  {'is_defined_by': 'Monarch',\n",
       "   'provided_by': ['https://data.monarchinitiative.org/ttl/omia.ttl'],\n",
       "   'evidence': [],\n",
       "   'qualifier': 'direct',\n",
       "   'publications': [],\n",
       "   'relation': 'RO:0002326',\n",
       "   'subject': 'NCBIGene:389756176',\n",
       "   'object': 'OMIA:001694-9940',\n",
       "   'id': 3269874,\n",
       "   'type': 'contributes_to',\n",
       "   'predicate': 'contributes_to'}),\n",
       " ('NCBIGene:200685841',\n",
       "  'OMIA:001089-9590',\n",
       "  {'is_defined_by': 'Monarch',\n",
       "   'provided_by': ['https://data.monarchinitiative.org/ttl/omia.ttl'],\n",
       "   'evidence': [],\n",
       "   'qualifier': 'direct',\n",
       "   'publications': [],\n",
       "   'relation': 'RO:0002326',\n",
       "   'subject': 'NCBIGene:200685841',\n",
       "   'object': 'OMIA:001089-9590',\n",
       "   'id': 3948919,\n",
       "   'type': 'contributes_to',\n",
       "   'predicate': 'contributes_to'}),\n",
       " ('NCBIGene:403511',\n",
       "  'OMIA:001244-9615',\n",
       "  {'is_defined_by': 'Monarch',\n",
       "   'provided_by': ['https://data.monarchinitiative.org/ttl/omia.ttl'],\n",
       "   'evidence': [],\n",
       "   'qualifier': 'direct',\n",
       "   'publications': [],\n",
       "   'relation': 'RO:0002326',\n",
       "   'subject': 'NCBIGene:403511',\n",
       "   'object': 'OMIA:001244-9615',\n",
       "   'id': 4024483,\n",
       "   'type': 'contributes_to',\n",
       "   'predicate': 'contributes_to'}),\n",
       " ('NCBIGene:403972',\n",
       "  'OMIA:001244-9615',\n",
       "  {'is_defined_by': 'Monarch',\n",
       "   'provided_by': ['https://data.monarchinitiative.org/ttl/omia.ttl'],\n",
       "   'evidence': [],\n",
       "   'qualifier': 'direct',\n",
       "   'publications': [],\n",
       "   'relation': 'RO:0002326',\n",
       "   'subject': 'NCBIGene:403972',\n",
       "   'object': 'OMIA:001244-9615',\n",
       "   'id': 3951300,\n",
       "   'type': 'contributes_to',\n",
       "   'predicate': 'contributes_to'}),\n",
       " ('NCBIGene:43056830',\n",
       "  'OMIA:001089-9598',\n",
       "  {'is_defined_by': 'Monarch',\n",
       "   'provided_by': ['https://data.monarchinitiative.org/ttl/omia.ttl'],\n",
       "   'evidence': [],\n",
       "   'qualifier': 'direct',\n",
       "   'publications': [],\n",
       "   'relation': 'RO:0002326',\n",
       "   'subject': 'NCBIGene:43056830',\n",
       "   'object': 'OMIA:001089-9598',\n",
       "   'id': 3750612,\n",
       "   'type': 'contributes_to',\n",
       "   'predicate': 'contributes_to'}),\n",
       " ('NCBIGene:38908820',\n",
       "  'OMIA:001450-9913',\n",
       "  {'is_defined_by': 'Monarch',\n",
       "   'provided_by': ['https://data.monarchinitiative.org/ttl/omia.ttl'],\n",
       "   'evidence': [],\n",
       "   'qualifier': 'direct',\n",
       "   'publications': [],\n",
       "   'relation': 'RO:0002326',\n",
       "   'subject': 'NCBIGene:38908820',\n",
       "   'object': 'OMIA:001450-9913',\n",
       "   'id': 4064365,\n",
       "   'type': 'contributes_to',\n",
       "   'predicate': 'contributes_to'}),\n",
       " ('NCBIGene:38908820',\n",
       "  'OMIA:001464-9913',\n",
       "  {'is_defined_by': 'Monarch',\n",
       "   'provided_by': ['https://data.monarchinitiative.org/ttl/omia.ttl'],\n",
       "   'evidence': [],\n",
       "   'qualifier': 'direct',\n",
       "   'publications': [],\n",
       "   'relation': 'RO:0002326',\n",
       "   'subject': 'NCBIGene:38908820',\n",
       "   'object': 'OMIA:001464-9913',\n",
       "   'id': 3830535,\n",
       "   'type': 'contributes_to',\n",
       "   'predicate': 'contributes_to'}),\n",
       " ('NCBIGene:199226050',\n",
       "  'OMIA:001089-9579',\n",
       "  {'is_defined_by': 'Monarch',\n",
       "   'provided_by': ['https://data.monarchinitiative.org/ttl/omia.ttl'],\n",
       "   'evidence': [],\n",
       "   'qualifier': 'direct',\n",
       "   'publications': [],\n",
       "   'relation': 'RO:0002326',\n",
       "   'subject': 'NCBIGene:199226050',\n",
       "   'object': 'OMIA:001089-9579',\n",
       "   'id': 4154152,\n",
       "   'type': 'contributes_to',\n",
       "   'predicate': 'contributes_to'})]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show the first 10 edges from the networkx graph\n",
    "monarch_lite_transformer.graph.edges(data=True)[0:9]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'monarch_lite_data.csv.tar'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a PandasTransformer, required for CSV export\n",
    "pandas_transformer = PandasTransformer(monarch_lite_transformer.graph)\n",
    "\n",
    "# export the networkx graph as CSV\n",
    "pandas_transformer.save(\"monarch_lite_data.csv\")\n",
    "\n",
    "# You should see monarch_lite_data.csv.tar in your current folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
